daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -0,0 +1,259 @@
+  import valueTranslator.unsafeTranslateValue
+
+  // This pulls all the dependencies of in `typesToProcess0` and `tyConAlreadySeed0`
+  private def getDependencies(
+      typesToProcess0: List[Ast.Type],
+      tmplToProcess0: List[Ref.TypeConName],
+      tyConAlreadySeed0: Set[Ref.TypeConName] = Set.empty,
+      tmplAlreadySeed0: Set[Ref.TypeConName] = Set.empty,
+  ): Result[(Set[Ref.TypeConName], Set[Ref.TypeConName])] = {
+
+    @tailrec
+    def go(
+        typesToProcess0: List[Ast.Type],
+        tmplToProcess0: List[Ref.TypeConName],
+        tyConAlreadySeed0: Set[Ref.TypeConName],
+        tmplsAlreadySeed0: Set[Ref.TypeConName],
+    ): Result[(Set[Ref.TypeConName], Set[Ref.TypeConName])] =
+      typesToProcess0 match {
+        case typ :: typesToProcess =>
+          typ match {
+            case Ast.TApp(fun, arg) =>
+              go(fun :: arg :: typesToProcess, tmplToProcess0, tyConAlreadySeed0, tmplsAlreadySeed0)
+            case Ast.TTyCon(tyCon @ Ref.Identifier(packageId, qualifiedName))
+                if !tyConAlreadySeed0(tyCon) =>
+              compiledPackages.packages.lift(packageId) match {
+                case Some(pkg) =>
+                  PackageLookup.lookupDataType(pkg, qualifiedName) match {
+                    case Right(Ast.DDataType(_, _, dataType)) =>
+                      val typesToProcess = dataType match {
+                        case Ast.DataRecord(fields, _) =>
+                          fields.foldRight(typesToProcess0)(_._2 :: _)
+                        case Ast.DataVariant(variants) =>
+                          variants.foldRight(typesToProcess0)(_._2 :: _)
+                        case Ast.DataEnum(_) =>
+                          typesToProcess0
+                      }
+                      go(
+                        typesToProcess,
+                        tmplToProcess0,
+                        tyConAlreadySeed0 + tyCon,
+                        tmplsAlreadySeed0)
+                    case Left(e) =>
+                      ResultError(e)
+                  }
+                case None =>
+                  ResultNeedPackage(
+                    packageId,
+                    _ =>
+                      getDependencies(
+                        typesToProcess0,
+                        tmplToProcess0,
+                        tyConAlreadySeed0,
+                        tmplsAlreadySeed0))
+              }
+            case Ast.TTyCon(_) | Ast.TNat(_) | Ast.TBuiltin(_) | Ast.TVar(_) =>
+              go(typesToProcess, tmplToProcess0, tyConAlreadySeed0, tmplsAlreadySeed0)
+            case Ast.TSynApp(_, _) | Ast.TForall(_, _) | Ast.TStruct(_) =>
+              ResultError(Error(s"unserializable type ${typ.pretty}"))
+          }
+        case Nil =>
+          tmplToProcess0 match {
+            case tmplId :: tmplsToProcess if tmplsAlreadySeed0(tmplId) =>
+              getDependencies(Nil, tmplsToProcess, tyConAlreadySeed0, tmplsAlreadySeed0)
```

Comment:
Should this refer to `go`?
```suggestion
              go(Nil, tmplsToProcess, tyConAlreadySeed0, tmplsAlreadySeed0)
```